### ISSUE TITLE
useCallback() invalidates too often in practice

### ISSUE THREAD

**Issue by gaearon** [facebook/react#14099]
This is related to https://github.com/facebook/react/issues/14092, https://github.com/facebook/react/issues/14066, and some other issues.

The problem is that we often want to avoid invalidating a callback (e.g. to preserve shallow equality below or to avoid re-subscriptions in the effects). But if it depends on props or state, it's likely it'll invalidate too often. See https://github.com/facebook/react/issues/14092#issuecomment-435907249 for current workarounds.

`useReducer` doesn't suffer from this because the reducer is evaluated directly in the render phase. @sebmarkbage had an idea about giving `useCallback` similar semantics but it'll likely require complex implementation work. Seems like we'd have to do _something_ like this though.

I'm filing this just to acknowledge the issue exists, and to track further work on this.

**Comment by gokalina** [comment]
@gaearon thank you for your answer in reactjs/rfcs#83. I've look at sources of `useReducer`. But I can't understand how it is related to `useCallback`. What issues has "mutation of ref during rendering"? Can you explain me in brief?

**Comment by gaearon** [comment]
> I've look at sources of useReducer. But I can't understand how it is related to useCallback

`useCallback` lets you memoize the callback to avoid a different function being passed down every time. But you have to specify everything it depends on in the second array argument. If it's something from props or state, your callback might get invalidated too often.

`useReducer` doesn't suffer from this issue. The `dispatch` function it gives you will stay the same between re-renders even if the reducer itself closes over props and state. This works because the reducer runs during the next render (and thus has natural ability to read props and state). It would be nice if `useCallback` could also do something like this but it's not clear how.

>What issues has "mutation of ref during rendering"? Can you explain me in brief?

In concurrent mode (not yet released), it would "remember" the last rendered version, which isn't great if we render different work-in-progress priorities. So it's not "async safe".

**Comment by strayiker** [comment]
Would be nice if second argument of `useCallback` was injected as dependencies to callback function.

```javascript
  function useCallback(cb, deps) => {
    lastDeps = deps; // save current deps deep in somewhere

    if (!cached) {
      cached = (...args) => cb(...lastDeps)(...args); // memoize that forevere
    }

    return cached; // never invalidates
  }

  const myCallback = useCallback(
    (state, props) => (arg1, arg2) => {},
    [state, props]
  );

```

**Comment by gaearon** [comment]
Yes, that's the approach from https://github.com/reactjs/rfcs/issues/83 and https://reactjs.org/docs/hooks-faq.html#how-to-read-an-often-changing-value-from-usecallback. We don't want it to be default because it's easier to introduce bugs in concurrent mode this way.

**Comment by sophiebits** [comment]
@sokra An alternate would be:

```js
function useEventCallback(fn) {
  let ref = useRef();
  useLayoutEffect(() => {
    ref.current = fn;
  });
  return useCallback(() => (0, ref.current)(), []);
}
```

This doesn't require the args like yours has. But again, you can't call this in the render phase and the use of mutation is dicey for concurrent.

**Comment by sokra** [comment]
@sophiebits That's clever and would have none of the problems with args, etc. It even doesn't require an dependencies list.

One nitpick: `return useCallback((...args) => (0, ref.current)(...args), []);` to pass along i. e. `event` argument.

**Comment by strayiker** [comment]
@sokra With this you will not be able to access to state and props updates inside a callback.
```javascript
const [state, setState] = useState(0);

const handleClick = useCallback((event) => {
  console.log(state); // always 0
});

return <button onClick={handleClick} />
```
So dependencies is required.

```javascript
function useEventCallback(fn, deps) {
  const fnRef = useRef(fn);
  const depsRef = useRef(deps);

  useLayoutEffect(() => {
    fnRef.current = fn;
  });

  return useCallback(() => (0, ref.current)(), []);
}
```

**Comment by sokra** [comment]
> @sokra With this you will not be able to access to state and props updates inside a callback.

I think with @sophiebits' approach this will work. The latest function is always copied into the ref and only a trampoline function is returned. This will make sure that the latest function is called, which has the latest state in context.

**Comment by muqg** [comment]
I recently made a duplicate issue and was asked to check this one. What I proposed there was very similar to @sophiebits' approach, but still looks a bit simpler to me:

```JavaScript
function useStatic(cb) {
  const callback = useRef(cb)
  callback.current = cb

  const mem = useRef((...args) => callback.current(...args))
  return mem.current
  
  // We could, of course still, use the useCallback hook instead of a second reference.
  // return useCallback((...args) => callback.current(...args), [])
  // Altough I think that the one above is better since it avoids the need to compare anything at all.
}
```

This way it is guaranteed to update where the hook is called since it does not directly use any side effect and instead it only updates a reference. I seems to me that it will be callable during the render phase and should not be dicey with concurrent mode (unless references don't meet these two conditions). Wouldn't this approach be a little better or am I missing something?

**Comment by gaearon** [comment]
@muqg In concurrent mode, last render doesn't necessarily mean "latest committed state". So a low-priority render with new props or state would overwrite a reference used by current event handler.

**Comment by strayiker** [comment]
If I understand the problem correctly...

What if useCallback will return a special callable object with two different states of an our callback function? The first is a `current` callback, that will changes on each render. The second is a `commited` callback, that changes within a commit phase.
By default call to that object will trigger the `current` value, so it can be called in render phase directly. 
But internally, React will pass the `commited` value to the dom, which prevents the callback from being modified until the next commit.

```javascript
function Callback(cb) {
  function callback(...args) {
    return callback.current(...args);
  }

  callback.commited = cb;
  callback.current = cb;
  callback.SPECIAL_MARKER_FOR_REACT = true;

  return callback;
}

function useCallback(cb) {
  const callback = useMemo(() => new Callback(cb), []);

  callback.current = cb;
  
  useLayoutEffect(() => {
    callback.commited = cb;
  });

  return callback;
}
```
```javascript
function Component(counter) {
  const handler = useCallback(() => {
    console.log(counter);
  });

  handler(); // call to handler.current

  // pass handler.commited to the dom
  return <button onClick={handler} />
}
```

**Comment by Volune** [comment]
I don't think there is a point in saving the "current", if you want to call it during rendering, just save it in advance out of the hook:
```
const handler = () => {/* do something*/};
const callback = useCallback(handler);
// I can call the current:
handler();
```
----
I personally don't see any benefits of the current `useCallback` implementation over the proposed `useEventCallback`, will it become the new implementation?
Also, can it warn when the callback is called during render in development mode?

**Comment by strayiker** [comment]
Concurrent mode can produce two different representations of a component (the first one is that commited to the dom and the second one that in memory). This representations should behaves accordingly with their props and state.

`useEventCallback` by @sophiebits mutates `ref.current` after all dom mutations is completed, so the `current` (in-memory) component can't use the newest callback.

@muqg proposal mutate the callback on each render, so the `commited` component will lose the reference to the `old` callback.

The point of my proposal in the passing a separated callback reference, that will changes in commit phase, to the dom, while the in-memory (not commited) representation of a component can use the latest version of that callback.

**Comment by Voronar** [comment]
Hi.
According to @sophiebits `useEventCallback` implementation why is it 
function uses `useLayoutEffect` and not `useEffect` for `ref` updating?
And is it normal due to current limitations use `useEventCallback` for all internal functions (which wants be memoized for 
using in expensive pure components tree and has closured variables from outer hook function) inside hook?

**Comment by jonnyasmar** [comment]
@Bazzer588 Are you using React.memo on your functional component? What do your attempts using hooks/functional look like? Your problem may or may not be related to this issue; it's hard to tell without seeing your code.

**Comment by Bazzer588** [comment]
Here's the full code - just drop a `<TestForm/>` or a `<HookTestForm/>` somewhere on a page

Using a React.Component, only the field being editted does a full render

```
import React from 'react';
import TextField from "./TextField";

export default class TestForm extends React.Component {

    onChangeField = (name,value) => {
        this.setState({ [name]: value });
    };

    render () {
        const state = this.state || {};
        return (
            <>
                <TextField name="gender"      value={state.gender}      onChangeField={this.onChangeField} />
                <TextField name="firstName"   value={state.firstName}   onChangeField={this.onChangeField} />
                <TextField name="lastName"    value={state.lastName}    onChangeField={this.onChangeField} />
                <TextField name="dateOfBirth" value={state.dateOfBirth} onChangeField={this.onChangeField} />
            </>
        );
    }
}
```

Using Hooks, all the child elements are re-rendered every time - presumably as onChangeField changes every time the state data changes. Is there some way I can implement onChangeField so it behaves like the above example?

```
import React, {useState, useCallback} from 'react';
import TextField from "./TextField";

export default React.memo( () => {

    const [data, changeData] = useState({});

    const onChangeField = useCallback((name,value) => {
        changeData({ ...data, [name]: value });
    }, [data] );

    return (
        <>
            <TextField name="gender"      value={data.gender}      onChangeField={onChangeField} />
            <TextField name="firstName"   value={data.firstName}   onChangeField={onChangeField} />
            <TextField name="lastName"    value={data.lastName}    onChangeField={onChangeField} />
            <TextField name="dateOfBirth" value={data.dateOfBirth} onChangeField={onChangeField} />
        </>
    );
});
```

This is my <TextField> component - you can see when it does a full render from the console or with the React dev tools

```
import React from 'react';

export default React.memo( ({ name, value, onChangeField }) => {

    console.log('RENDER TEXT FIELD',name,value);

    return (
        <div className="form-field">
            <label htmlFor={name}>{name}</label>
            <input
                type="text"
                onChange={ (ev) => onChangeField( name, ev.target.value ) }
                value={value || ''}
            />
        </div>
    );
});

```

**Comment by muqg** [comment]
@Bazzer588 I think its due to the object value kept in state under the variable name of `data`. I don't know why but objects in state always invalidate memoization and  thus your callback `onChangeField` is a new on on each render thus breaking the memoization of the components you're passing it to.

I've had a similar issue like you and noticed this as being its cause. I have no idea why the object in state does not keep its reference when it has not been explicitly set to a new object.

**Comment by Bazzer588** [comment]
Yes my problem is that the in first example (using React.Component) I can create a _onChangeField_ callback which is bound to _this_, and never changes during renders

Using the new hook methods I can't seem to replicate the way the existing functionality works, 

On the project I'm working on we often do this to have a hierachy of components and state:

```
    onChangeField = (fieldName,fieldValue) => {
        const newValue = { ...this.props.value, [fieldName]: fieldValue };
        this.props.onChangeField( this.props.name, newValue );
    };
```
So it passes props down the tree (ie value {})
And uses the callback to send new values up the tree

**Comment by jonnyasmar** [comment]
@Bazzer588 The data var your passing to the second parameter of useCallback is going to invalidate every time. useCallback doesn't do a deep comparison. You need to pass in a flat array for it to properly memoize.

**Comment by Bazzer588** [comment]
Yes I've tried similar with _useReducer_ and passing _dispatch_ down to the child components - this does seem to work as you get the same _dispatch_ method so it doesn't force a re-render

If there's no way to make callback functions work as they did before I guess that's the way we'll have to go

**Comment by sebmarkbage** [comment]
@Bazzer588 The general recommendation is to `useReducer`.

That said, your particular example can be solved like this but it is pretty unusual to not depend on any props so this doesn't always work:

```js
    const onChangeField = useCallback((name, value) => {
        changeData(oldData => ({ ...oldData, [name]: value }));
    }, []);
```

**Comment by Bazzer588** [comment]
Thanks @sebmarkbage - didn't realise you could pass a function to changeData
(As in `const [data, changeData] = useState({});` )
However now I'm worried - in what case would this _doesn't always work_ ?

**Comment by doasync** [comment]
@gaearon, is this issue going to be solved before hooks release?

**Comment by gaearon** [comment]
Workarounds above (especially relying more on `useReducer`) seem sufficient for most cases. There are cases when they’re not, but we’ll likely revisit this again in a few months.

**Comment by Bazzer588** [comment]
It's very, very different behaviour

With a ES6 class I can pass _onChange={this.handleChange}_ to the child component and it does not redraw every time (because it sends the same _handleChange_ method every time)

It's frustrating that _dispatch_ works independent of state (I mean the _dispatch_ method does not change even if the underlying state does) but useCallback does not (if your callback needs to do something with state that is)

I see a lot of code where the whole form redraws when the use types into a field

A lot of devs don't understand this until users complain 'it's slow'

**Comment by doasync** [comment]
So, why do we need useCallback after all? It can be replaced with:
```js
const useCallback = fn => useReducer(fn)[1];
```
just because:

> The `dispatch` function it gives you will stay the same between re-renders even if the reducer itself closes over props and state. This works because the reducer runs during the next render (and thus has natural ability to read props and state).

We also don't need to specify dependencies in the second argument

**Comment by HaNdTriX** [comment]
I worked around this issue, by creating the following hook:

**Code:**

```js
import { useRef, useEffect } from 'react'

/**
 * This hook reduces the amount of rebindings.
 * Use it when your props or state change a lot 
 * or when you have nested handlers.
 *
 * @param {DomRef} targetRef The reference to the dom node
 * @param {String} eventName The eventName you want to bind
 * @param {Function} handler The actual event handler
 */
function useCachedRefHandler (targetRef, eventName, handler) {
  const savedHandler = useRef()

  useEffect(() => {
    savedHandler.current = handler
  }, [handler])

  useEffect(() => {
    const handleAttach = (...args) => {
      if (savedHandler.current) {
        savedHandler.current(...args)
      }
    }
    targetRef.current.addEventListener(eventName, handleAttach)
    return () => targetRef.current.removeEventListener(eventName, handleAttach)
  }, [])
}

export default useCachedRefHandler
```

This code allows me to omit `useCallback` completely and therefore use state and props without diffing. 

**Usage Example:**

```js
import React, { useRef } from 'react'
import useCachedRefHandler from './hooks/useCachedRefHandler'

function Webview({ onSomeingA, onSomethingB, onSomethingC }) {
  const ref = useRef()

  useCachedRefHandler(ref, 'ipc-message', (event) => {
    switch (event.channel) {
      case 'someingA':
        onSomeingA (event.data)
        return
      case 'someingB':
        onSomeingB (event.data)
        return
      case 'someingC':
        onSomeingB (event.data)
        return
    }
  })

  return (
    <webview ref={ref} />
  )
}

export default Webview
```

**Comment by williamoliveira** [comment]
I was wondering what happens if I do side effects inside setState's callback

```js
const handler = useCallback(() => {
    setState((state) => {
        doSomethingThatDependsOnState(state)
        return state
    })
}, [])
```

How bad is it? will babies die or something?

**Comment by eps1lon** [comment]
Had a use case for @sophiebits proposed `useEventCallback` (https://github.com/facebook/react/issues/14099#issuecomment-440013892). The problem with `useLayoutEffect` is that it issues warnings when server side rendering. You can trick React by switching to `useEffect` based on the environment (`typeof window`) but this won't work if the SSR API is called in the browser (`with-apollo`).

Wouldn't `useImperativeHandle` work as well?


```diff
function useEventCallback(fn) {
  let ref = useRef();
-  useLayoutEffect(() => {
+  useImperativeHandle(ref, () => {
-    ref.current = fn;
+    return fn;
  });
  return useCallback(() => (0, ref.current)(), []);
}
```

**Comment by sompylasar** [comment]
I found myself doing this often:

```jsx
function Foo({ someProp, otherProp }) {
  const somePropRef = useRef(someProp);
  useLayoutEffect(() => {  // or useEffect
    somePropRef.current = someProp;
  }, [someProp]);

  const otherPropRef = useRef(otherProp);
  useLayoutEffect(() => {  // or useEffect
    otherPropRef.current = otherProp;
  }, [otherProp]);

  const onSomething = useCallback(() => {
    handleSomething(somePropRef.current, otherPropRef.current);
  }, []);

  return <Bar onSomething={onSomething} />;

  // or...
  //
  // // This ref value never changes, so can use .current in render.
  // // WARNING: Safe to close over refs and outer scope consts only.
  // const onSomethingRef = useRef(() => {
  //   handleSomething(somePropRef.current, otherPropRef.current);
  // });
  //
  // return <Bar onSomething={onSomethingRef.current} />;
}
```

Does that have any drawbacks besides unnecessary effects?

**Comment by arnotes** [comment]
> I found myself doing this often:
> 
> ```js-jsx
> function Foo({ someProp, otherProp }) {
>   const somePropRef = useRef(someProp);
>   useLayoutEffect(() => {  // or useEffect
>     somePropRef.current = someProp;
>   }, [someProp]);
> 
>   const otherPropRef = useRef(otherProp);
>   useLayoutEffect(() => {  // or useEffect
>     otherPropRef.current = otherProp;
>   }, [otherProp]);
> 
>   const onSomething = useCallback(() => {
>     handleSomething(somePropRef.current, otherPropRef.current);
>   }, []);
> 
>   return <Bar onSomething={onSomething} />;
> 
>   // or...
>   //
>   // // This ref value never changes, so can use .current in render.
>   // // WARNING: Safe to close over refs and outer scope consts only.
>   // const onSomethingRef = useRef(() => {
>   //   handleSomething(somePropRef.current, otherPropRef.current);
>   // });
>   //
>   // return <Bar onSomething={onSomethingRef.current} />;
> }
> ```
> 
> Does that have any drawbacks besides unnecessary effects?

@sompylasar I wanna know as well. we are doing the same thing using this hook
```
useController<T extends IPropsAndState>(propsAndState:T){
  const controllerRef = useRef({});
  Object.assign(controllerRef.current, propsAndState);
  return controllerRef.current;
}
....

....


```

we badly need it since we have a websocket rjxs subscription and we need an updated context for the callback functions inside filter pipes and the actual subscription.

i cant use a useeffect that contains dependencies in my case because i dont want to constantly disconnect/reconnect to our socket server

**Comment by nikparo** [comment]
@arnotes Looks like you are updating your ref during the render phase, i.e. before it has been commited, rather than in a `useEffect` call. To quote @gaearon : 

> In concurrent mode, last render doesn't necessarily mean "latest committed state". So a low-priority render with new props or state would overwrite a reference used by current event handler.

Afaik the drawback with updating the ref in a `useEffect` is mainly that you then can't use it during the render.

**Comment by bvaughn** [comment]
> Does that have any drawbacks besides unnecessary effects?

Late to the party here, but after scanning the above thread- there is another potential problem in [the proposed `useEventCallback`](https://github.com/facebook/react/issues/14099#issuecomment-440013892) and other example uses of `useLayoutEffect` above:

```js
function Example({ callback }) {
  const stableCallback = useEventCallback(callback);
  return <Child callback={stableCallback} />;
}

function Child({ callback }) {
  useLayoutEffect(() => {
    // Callback has not yet been updated!
    // This is because child effects are run before parent effects,
    // So the callback ref has not yet been updated (and still points to the old value).
    callback();
  }, [callback]);
}

```

**Comment by yelouafi** [comment]
>>What issues has "mutation of ref during rendering"? Can you explain me in brief?

>In concurrent mode (not yet released), it would "remember" the last rendered version, which isn't great if we render different work-in-progress priorities. So it's not "async safe".

Aren't class components also exposed to this? Seems React does the same thing to Class Components (mutating props in render phase)

https://github.com/facebook/react/blob/4c5698400f04bbc6d0b4bd766b0993d0bcb37609/packages/react-reconciler/src/ReactFiberClassComponent.js#L1042-L1045

als seems the committed and work-in-progress fibers share the same class instance

https://github.com/facebook/react/blob/4c5698400f04bbc6d0b4bd766b0993d0bcb37609/packages/react-reconciler/src/ReactFiber.js#L413

If React assigns pending props to `instance.props` in the render phase but the render work is interrupted for some reason, now `this.props` reflects the props of the WIP. But if an event handler fires in between, it should access the committed props.

I assume the solution is to switch `this.props` between phases, maybe this is already done but I couldn't trace it in the sources so just wanted to make sure this is handled.

**Comment by vzaidman** [comment]
I like the following syntax:
```
  const increaseCount = useSmartCallback(
    ({ setCount, count }) => setCount(count + 1),
    { setCount, count }
  );
```
where `increaseCount` is never re-created, but always gets passed with new props.

with the right eslint that makes sure no props from the upper scope are used inside of the `cb` of `useSmartCallback`, it might be very useful.

this hook might be look like this:
```
const useSmartCallback = (cb, deps) => {
  const ref = React.useRef(null);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  React.useEffect(() => ref.current = deps, Object.values(deps));

  // eslint-disable-next-line react-hooks/exhaustive-deps
  return React.useCallback(() => cb(ref.current), []);
};
```
I created a sandbox that proves it:
https://codesandbox.io/s/usesmartcallback-nts9k

**Comment by Hypnosphi** [comment]
@vzaidman

The [solution](https://github.com/facebook/react/issues/14099#issuecomment-440013892) by @sophiebits does the same without a need for stating any deps at all

**Comment by vzaidman** [comment]
missed it altogether and deleted my comment. thanks

**Comment by Hypnosphi** [comment]
@Volune 

> I personally don't see any benefits of the current `useCallback` implementation over the proposed `useEventCallback`

Current `useCallback` is quite helpful with render props pattern. One example is `List` from `react-virtualized`. It's a `PureComponent`, so you can pass a memoized `rowRenderer` function to prevent rerenders. But if your `rowRenderer` doesn't invalidate when its deps change (e.g. if it's an instance property on a class component), you have to pass all those deps as `List` props to trigger rerenders when they're actually needed.

Example on codesandbox: https://codesandbox.io/embed/staging-snow-sgh7v
Note how I have to pass an extra `value` property to `List` to fix the class example while `useCallback` works as is because it invalidates whenever it would produce a different output.

@muqg @strayiker BTW this is why there's no much sense in trying to make `useEventCallback` work in the render phase: If a child uses your callback to render something, you most likely _want_ it to invalidate as often as its deps change. Also, this is why `useEventCallback` is a really good name. Basically, it says "only use me for event handlers".

**Comment by artem-malko** [comment]
@gaearon @sebmarkbage

Lets imagen, we have react-redux in our App. So, in any component, where we are going to dispatch something we will code something like this:

```jsx
const Comp = React.memo((props) => {
  // This is from react-redux
  const dispatch = useDispatch();
  const handler = useCallback(() => {
    // actionCreator is created in our project and it is static function
    dispatch(actionCreator(props));
  }, [props]);

  return <Button onClick={handler}></Button>
})
```
So,  Button component will receive new handler every time, than props are changed. So, as I understand, this is the main problem of current issue)

What if we will create our own custom hook useHandler, which will look like:
```jsx
// actionCreator won't be changed, just static function in our case
const useHandler = (actionCreator, params) => {
  // This is from react-redux. Dispatch won't be changed.
  const dispatch = useDispatch();
  const paramsRef = useRef(params);
  // This is useless for the first time, but we need it later
  // This code can be moved or we can add condition that paramsRef.current !== params.
  paramsRef.current = params;
  // callback will be created only on the first call of useHandler, cause of []
  // But it will dispatch action with new params on next calls
  const callback = useCallback(() => {
    dispatch(actionCreator(paramsRef.current));
  }, []);
  
  return callback;
}
```
So, if we have static actionCreator, static dispatch from useDispatch — is it safe to use that approach?

**Comment by bvaughn** [comment]
This line in your above example could cause problems:
```js
paramsRef.current = params;
```

Since you're mutating a ref during render- and renders can be async- you can end up mutating it prematurely (so that a previously-rendered thing calls the `callback` and gets a not-yet-rendered `params` value.

To avoid this, you would want to update the ref in a layout effect.

**Comment by artem-malko** [comment]
@bvaughn but in that case it is possible to have a bug like in your example here https://github.com/facebook/react/issues/14099#issuecomment-534673602

**Comment by bvaughn** [comment]
Yup! Still has potential for bugs. Mutating during during render seems worse to me though.

**Comment by artem-malko** [comment]
As I understand, we can use safely a value from useRef in any hook, but we can not change it safely. Only React itself can do it. For example, if we use useRef to have a ref on dom-node — it is ok, cause React change current value.

**Comment by artem-malko** [comment]
@bvaughn 
Ok, thanks. Now it's much more clear for me) useEffect is not a silver bullet, but a compromise for useHandler in case, when fn for callback is just static function and only params are changing.

So, will wait for new hook, which is based on useReducer or something like this.

**Comment by igodorogea** [comment]
@sophiebits Why is the comma expression needed here?

> ```js
> useEventCallback(fn) {
>   ...
>   return useCallback(() => (0, ref.current)(), []);
> }
> ```

**Comment by ljharb** [comment]
It ensures that the receiver (this value) of the function isn’t ref.

**Comment by ypresto** [comment]
`useImperativeHandle()` solution looks good, but it turns out that ref.current will not be set when the callback called in `useLayoutEffect()` placed in a child component.

https://codesandbox.io/s/react-useeventcallback-test-eirgj

**Comment by ypresto** [comment]
@eps1lon 
https://github.com/facebook/react/issues/14099#issuecomment-499781277

The `useImperativeHandle()` solution looks good, but it turns out that ref.current will not be set when the callback called in `useLayoutEffect()` placed in a child component.

https://codesandbox.io/s/react-useeventcallback-test-eirgj
This PoC code shows that ref.current is always not set at callback call.

**Comment by ypresto** [comment]
The `useImperativeHandle()` solution still does not call outdated func at least. So you can use it where you are confident that the callback will not be called from useLayoutEffect().

**Comment by vzaidman** [comment]
Just to make sure-
Does binding functions in class components to avoid any invalidation of a callback we pass would work fine in concurrent mode?

f.e:
```
class Parent extends React.Component {
  state = {count: 0}

  handleClick = () => {
    const {count} = this.state;
    console.log(`hi #${count}!`);
  }
```

> @muqg In concurrent mode, last render doesn't necessarily mean "latest committed state". So a low-priority render with new props or state would overwrite a reference used by current event handler.

**Comment by jedwards1211** [comment]
Tbh this is making me super skeptical of concurrent mode, given that function memoization is the number one tedious thing I deal with daily in React since hooks were introduced (everything else about hooks being fantastic) and concurrent mode wouldn't benefit me personally right now AFAIK.

**Comment by mariusGundersen** [comment]
I came across this problem when I tried to call one callback from another one, and observed strange effects. A very simplified example:

```
const MyComponent = props=> {
  const [value, setValue] = useState();
  const [copy, setCopy] = useState();

  const createCopy = useCallback(() => {
    setCopy(value);
  }, [value]);

  const callback = useCallback(() => {
    createCopy();
  }, []);

  return (
    <div>
      <input onChange={e => setValue(e.target.value)} />
      <button onClick={callback}>Copy</button>
      <span>{copy}</span>
    </div>
  );
};
```

This is obviously a simplified and therefore contrived example. It does not behave how I would expect it to.

According to [the documentation](https://reactjs.org/docs/hooks-reference.html#usestate) it is safe to use `setCopy` from within a callback (or effect) without adding it to the dependency list. But it appears it's not safe to call another callback created with `useCallback` without adding it to the dependency list. There is probably a good reason for it, but it seems weird to me and clearly several other people in this thread.

So in the above example calling `callback()` will not update `copy`, since it calls an old version of `createCopy` which holds an old version of `value`.

**Comment by Hypnosphi** [comment]
@mariusGundersen `setCopy` is guaranteed to remain the same through component's lifetime, while `createCopy` changes each time when `value` does

**Comment by mariusGundersen** [comment]
Yes, that's clearly how it behaves. But it also means that `callback` needs to change too. It would be useful and consistent if `useCallback` behaved like `useState`.

**Comment by davidnussio** [comment]
@mariusGundersen If you want callback change too, it need setCopy as its dependency.
In this way it is guarantee that no side effects happened

**Comment by liorJuice** [comment]
I found that you can use useReducer to mitigate this effectively. the bit that was hard to swallow is when i needed side effects to happen in the callback.

Side effects **cant** happen in the reducer as it is being called with past actions _a lot_, on re renders, for some reason. so the reducer must be pure (state, action) => resultState, with no side effects.

The solution was to use useEffect that listen to changes in the reduced state (resultState), to initiate side effects, which suddenly seemed obvious. 😅

for example, this code:

```
  const [counter, setCounter] = useState(0);

  const handleClick = useCallback(() => {
    console.log(`child clicked ${counter + 1} times`);
    setCounter((c) => c + 1);
  }, [counter]);
```

which will create a very problematic callback, could be perfectly replaced with:

```
  const [counter, handleClick] = useReducer((c) => c + 1, 0);

  useEffect(() => {
    counter && console.log(`child clicked ${counter} times`);
  }, [counter]);
```

which will have the same effect but with handler that will persist through the entire lifetime of the component.

i think this is the most complete solution to the useCallback problem.
@gaearon can you confirm?

**Comment by liorJuice** [comment]
@Hypnosphi like i use the counter, as a useEffect dependency.. you can escape effect triggers by their changes if you want.

**Comment by Hypnosphi** [comment]
@liorJuice

> like other effect dependencies

then the effect will trigger when the prop change which is probably not what you want.

I'm talking about cases like that:

```
const handleClick = useCallback(() => {
  console.log(props.name);
}, [props.name]);
```

How would you apply your approach to this case?

**Comment by jedwards1211** [comment]
You wouldn't need to pass props used in side effects as dependencies -- the new values for those props would be part of the new closure when the effect fires because the counter dependency changed.

```
  useEffect(() => {
    counter && console.log(`${props.name}: child clicked ${counter} times`);
  }, [counter]);
```

Though I assume it may be possible for the props to change between the click that incremented the counter and the render that triggers the effect?

**Comment by liorJuice** [comment]
@Hypnosphi the problem you describe is a general useEffect problem (how to write effects with dependencies that don't need to trigger the effect),
it is not specific to the the useCallback solution i described.
there are number of approaches you can take to solve prop dependencies in effects.
i suggest taking a look here:
https://overreacted.io/a-complete-guide-to-useeffect/

and also search for a custom hook named "usePrevious", you can use it to check if a specific dependency changed on useEffect.

**Comment by nikparo** [comment]
I was inspired by @liorJuice's [thoughts on useReducer -> useEffect above](https://github.com/facebook/react/issues/14099#issuecomment-641811678) to create a custom hook as an experiment, [use-effect-callbacks](https://github.com/nikparo/use-effect-callbacks/). It transforms an object of callbacks into stable methods that can then be passed down. It works by caching the callbacks through a `useReducer` and executing them in the next `useEffect`. Could be used as such:

```js
import { useEffectCallbacks } from 'use-effect-callbacks';

// ...

// Within parent component
const [state, setState] = useState({ foo: 'bar' });
const { getState } = useEffectCallbacks({
  getState: () => state,
});

// ...

// Within memoized child component
const handleClick = async () => {
  setState({ foo: 'baz' });
  const latestState = await getState();
  console.log(latestState); // -> { foo: 'baz' }
};
```

#### Pros:
- It should be concurrent safe.
- The methods are stable as long as the callback names are stable.
- It is quite straightforward and simple to use.
- You don't need to manually keep track of a lot of different properties in your state to trigger various effects. (Compared to the more manual approach proposed by @liorJuice.)

#### Cons:
- It's a bit wasteful.
  - Calling a method will trigger the owner component to re-render. (Not a problem if you expect it to re-render anyway, e.g. due to a setState call)
  - If a method is called in a `useEffect`, I believe it will still trigger a full render-cycle and only execute in the next `useEffect`.
- The methods are not suitable as event listeners since they execute later, and any react event would have been recycled. (They can however very well be called from event listeners)
- While the methods can "safely" be called during renders, there's not much point in doing so since they return promises.

To be perfectly honest I'm not quite convinced. The approach seems handy for occasionally getting the **next** state, but most other use-cases I can think of seem like a bit of a stretch.

**Comment by ypresto** [comment]
FYI: My snippet of useImperativeHandle solution https://github.com/facebook/react/issues/14099#issuecomment-569044797 is like this:

```ts
// Better useCallback() which always returns the same (wrapped) function reference and does not require deps array.
// Use this when the callback requires to be same ref across rendering (for performance) but parent could pass a callback without useCallback().
// useEventCallback(): https://github.com/facebook/react/issues/14099#issuecomment-499781277
// WARNING: Returned callback should not be called from useLayoutEffect(). https://github.com/facebook/react/issues/14099#issuecomment-569044797
export function useStableCallback<T extends (...args: any[]) => any>(fn: T): T {
  const ref = useRef<T>()
  useImperativeHandle(ref, () => fn) // Assign fn to ref.current (currentFunc) in async-safe way

  return useRef(((...args: any[]) => {
    const currentFunc = ref.current
    if (!currentFunc) {
      throw new Error('Callback retrieved from useStableCallback() cannot be called from useLayoutEffect().')
    }
    return currentFunc(...args)
  }) as T).current
}

```

**Comment by dcheung2** [comment]
my workaround try to be KISS by abuse self reference from JS context.

But it always point to latest, not sure what could it break in various mode and lifecycle in react.
At least it work in my case.

```
export default function useStableCallback<T extends (...args: any[]) => unknown>(callback: T): T {
  const ref = useRef({
    stableProxy: (...args: any) => {
      return ref.current.targetFunc(...args);
    },
    targetFunc: callback,
  });
  ref.current.targetFunc = callback;
  return ref.current.stableProxy as T;
}
```

**Comment by Gentlee** [comment]
I got **problem** with **invalidated callback** when try to call it **after some state mutation**.

Very simplified example:

```javascript
const onClick = (value) => {
    setValue(value);
    navigateToNextPage();
}
```

Here `navigateToNextPage` uses old state in its closure and leads to a wrong screen.

I had to create `useNextUpdateCallback` hook to make it work on the next render:

```javascript
const navigateOnNextUpdate = useNextUpdateCallback(navigateToNextPage);

const onClick = () => {
    setValue(value);
    navigateOnNextUpdate();
}
```

**Implementation:**

```javascript
const forceUpdateReducer = (x) => x + 1;

export const useForceUpdate = () => {
  const [, forceUpdate] = useReducer(forceUpdateReducer, 0);
  return forceUpdate;
};

export const useNextUpdateCallback = (callback) => {
  // eslint-disable-next-line no-unused-vars
  const forceUpdate = useForceUpdate();
  const shouldCallRef = useRef(false); // ref is used not to rerender after callback is called

  useEffect(() => {
    if (!shouldCallRef.current) return;

    callback();
    shouldCallRef.current = false;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [shouldCallRef.current]);

  return useCallback(() => {
    shouldCallRef.current = true;
    forceUpdate();
  }, [forceUpdate]);
};
```

**Problems:**

- Navigation happens not immediately, this feels a little slower.
- It is a hack.

Does anyone know a better workaround?

PS. The more I use React and hooks, the more i understand how complex it is. Too complex.

**Comment by vdgodse** [comment]
I ran into a different problem but the workaround should help in your case as well. 
Let's assume **navigateToNextPage** depends on state **currentPage** in the closure. You could modify navigateToNextPage to read currentPage from an argument rather than the state in the closure.

> ```js
> 
> const onClick = () => {
>     setValue(value);
>     setCurrentPage((currentPage)=>{
>          navigateToNextPage(currentPage)
>          return currentPage
>     });
> }
> ```
> 

The drawback is navigateToNextPage is now inside an callback which is called **asynchronously**

**Comment by jedwards1211** [comment]
Hoping there will be some solution to this in React 18.

What if we had some kind of ref hook that safely keeps track of the last commited value on any given render thread?

```js
function useEventCallback(fn) {
  let ref = React.useConcurrentRef(fn);
  // value could be updated like:
  //   ref.setCurrent(fn);
  return useCallback((...args) => (0, ref.getCurrent)(...args), []);
}
```

Maybe it could have a `current` property with getters/setters just like `useRef`; I just used `setCurrent`/`getCurrent` in my example to illustrate that magic would be happening under the hood.

**Comment by windmaomao** [comment]
I smell useRef here. Either you want to manage the update or not.

**Comment by jampy** [comment]
Might [this approach](https://github.com/facebook/react/issues/14099#issuecomment-440717985) work fine when combined with [useConcurrentModeSafeRef()](https://dev.to/uhyo_/a-concurrent-mode-safe-version-of-useref-1325)?

**Comment by nikparo** [comment]
> Might [this approach](https://github.com/facebook/react/issues/14099#issuecomment-440717985) work fine when combined with [useConcurrentModeSafeRef()](https://dev.to/uhyo_/a-concurrent-mode-safe-version-of-useref-1325)?

`useConcurrentModeSafeRef` looks pretty cool at a first glance, but I'd want to know a bit more about this core line before using it in production:

```js
Promise.resolve().then(()=> raw.isRendering = false)
```

1. Do renders always complete during the same microtask? Including all child components? 
    - Can we count on this to remain true for concurrent renderering?
2. At exactly what point does the promise resolve? Before or after (layout) effects?

**Comment by jampy** [comment]
>useConcurrentModeSafeRef looks pretty cool at a first glance, but I'd want to know a bit more about this core line before using it in production:

Please check [this comment](https://dev.to/reaysawa/comment/10aap) at the end of the article, where João Paulo expresses similar thoughts and proposes to change that part to:

```
  raw.isRendering = true;
  raw.currentValue = raw.comittedValue;
  useLayoutEffect(() => {
    raw.isRendering = false;
    raw.comittedValue = raw.currentValue;
  });
```

The full implementation for `useConcurrentModeSafeRef()` would be like so, I guess (including a fix for the "committedValue" typo):

```tsx
type Raw<T> = {
  isRendering: boolean;
  committedValue: T;
  currentValue: T;
  ref: { current: T };
};
export const useConcurrentModeSafeRef = <T>(initialValue: T) => {
  const rawRef = useRef<Raw<T>>();
  const raw: Raw<T> = rawRef.current ?? (
    rawRef.current ={
      isRendering: true,
      committedValue: initialValue,
      currentValue: initialValue,
      ref: {
        get current() {
          if (raw.isRendering) {
            return raw.currentValue;
          } else {
            return raw.committedValue;
          }
        },
        set current(v) {
          if (!raw.isRendering) {
           raw.committedValue= v;
          }
          raw.currentValue = v;
        }
      }
    }
  );

  raw.isRendering = true;
  raw.currentValue = raw.committedValue;
  useLayoutEffect(() => {
    raw.isRendering = false;
    raw.committedValue= raw.currentValue;
  });

  return raw.ref;
};
```

So...

```js
function useStatic(cb) {
  const callback = useRef(cb)
  callback.current = cb

  const mem = useConcurrentModeSafeRef((...args) => callback.current(...args))
  return mem.current
}
```

This looks promising to me, but I'm not familiar enough about the lifecycle details (especially if the `useLayoutEffect()` callback fires before or after rendering children). Also, it's unclear to me if this covers all edge cases.

Probably only someone from the core team, like @gaearon, might be able to answer this.

*BTW, also have a look at [useMenoizedFn()](https://ahooks.js.org/hooks/use-memoized-fn/) ([source](https://github.com/alibaba/hooks/blob/3163809184594b03cfa3551e6698436492a112ea/packages/hooks/src/useMemoizedFn/index.ts)), which is a implementation like `useStatic()` above (ignoring the concurrent mode problem, IMHO, but avoiding [some problem when devtools is open](https://github-com.translate.goog/alibaba/hooks/issues/728?_x_tr_sl=zh-CN&_x_tr_tl=en&_x_tr_hl=de)).*

**Comment by nikparo** [comment]
> This looks promising to me, but I'm not familiar enough about the lifecycle details (especially if the useLayoutEffect() callback fires before or after rendering children). Also, it's unclear to me if this covers all edge cases.

Effects only run once a component is committed (a separate step to update the DOM after a component has been successfully rendered). But due to concurrent features such as suspense, there is no guarantee that a component that has started rendering will finish rendering, much less be committed.

In other words, `useLayoutEffect` is not a reliable way to tell whether it is still rendering. Afaik the component will eventually re-render and be committed once it's no longer held up, but in the meantime the hook will think that it is still rendering even though it isn't.

**Comment by jampy** [comment]
> Effects only run once a component is committed (a separate step to update the DOM after a component has been successfully rendered). But due to concurrent features such as suspense, there is no guarantee that a component that has started rendering will finish rendering, much less be committed.

I was aware of that. I actually meant "before or after *committing* children". Sorry for the confusion.

> In other words, useLayoutEffect is not a reliable way to tell whether it is still rendering.

Agree.

Another solution, that would work better, I think:

```jsx
function useConcurrentModeSafeRef(initialValue) {

  const isRendering = () =>
    !!React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner.current;

  const rawRef = useRef();

  const raw = rawRef.current ? rawRef.current : (
    rawRef.current = {
      comittedValue: initialValue,
      currentValue: initialValue,
      ref: {
        get current() {
          if (isRendering()) {
            return raw.currentValue;
          } else {
            return raw.comittedValue;
          }
        },
        set current(v) {
          if (!isRendering()) {
            raw.comittedValue = v;
          }
          raw.currentValue = v;
        }
      }
    }
  );

  raw.currentValue = raw.comittedValue;

  useLayoutEffect(() => {
    raw.comittedValue = raw.currentValue;
  });

  return raw.ref;

}

function useStatic(cb) {
  const callback = useRef(cb)
  callback.current = cb

  const mem = useConcurrentModeSafeRef((...args) => callback.current(...args))
  return mem.current
}
```

Of course, that's bad because it's using unsafe API (`React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED`). Perhaps something like `useConcurrentModeSafeRef()` could be added to React itself.

**Comment by Runeii** [comment]
> We've submitted a proposal to solve this. Would appreciate your input!
> 
> [reactjs/rfcs#220](https://github.com/reactjs/rfcs/pull/220)

This is great, but my immediate thought was "the useEvent name is a bit confusing". Is there a reason this couldn't be something more obvious like `useFunction`? `useCallback` always took some initial explanation to newcomers, though it made sense.

**Comment by mariusbrataas** [comment]
I wrote a hook that achieves exactly the same some time ago.

Would you please have a look, @gaearon? It might be every so slightly more efficient than yours, as it uses two hooks rather than three.

If you're interested, I also tried to start a discussion [here.](https://www.reddit.com/r/reactjs/comments/tsysgp/ref_callbacks_use_cases_and_pitfalls_of_this/)


```ts
/**
 * Returns a memoized version of your function that maintains a stable reference, but
 * also can read the latest scope (props and state) of the component in which it is used.
 */
function useStableCallback<Args extends any[], T>(
    callback: (...args: Args[]) => T
) {
    const callbackContainer = useRef(callback)
    callbackContainer.current = callback
    return useCallback(
        (...args: Args) => callbackContainer.current(...args),
        [callbackContainer]
    )
}
```

**Comment by bvaughn** [comment]
> const callbackContainer = useRef(callback)
> callbackContainer.current = callback

@mariusbrataas Updating a ref during render is an [unsafe side effect](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects).

**Comment by floroz** [comment]
The `useEvent` proposal definitely goes in the right direction, as the correct usage of `useCallback` and the potential pitfalls is something we've all seen countless times across teams and organizations. 

I do want to express some strong concerns around the naming as `useEvent` could potentially be very confusing for less experienced engineers or those who are coming to React from other frameworks.
The dangerous naming correlation to DOM events makes it quite misleading at first glance and seems to hide its true purpose.

I much prefer `useStableCallback` as someone else suggested, or `useHandler` as I feel it makes it easier to document and explain its usage moving forward.

**Comment by zheeeng** [comment]
There is a situation we shouldn't use `useEvent`:

```
const handleChangeValue = useCallback(
  (index) => (indexValue) => {
     onChange(listValue.map((v, i) => i === index ? indexValue : v))
  },
  [onChange, value]
)
```

useEvent keeps the reference of the raw value instead of the latest value.

**Comment by thmsobrmlr** [comment]
Definitely agree with @floroz regarding the naming and would prefer something like `useStableCallback`.

**Comment by Hypnosphi** [comment]
@zheeeng I'm not sure what's the benefit of even using `useCallback` in your case. `handleChangeValueByIndex(i)` will always return a new function even though the `handleChangeValueByIndex` itself is memoized

**Comment by zheeeng** [comment]
Hi, @Hypnosphi, I update this case here, the old style I use `useCallback` like this below. After switching to `useEvent` causes buggy behavior.

```
const handleChangeValueByIndex = useCallback(
  (index) => (indexValue) => {
     onChange(listValue.map((v, i) => i === index ? indexValue : v))
  },
  [onChange, value]
)

return (
   <>
      {value.map((v, i) => (
          <Item key={value.id} onChangeByIndex={handleChangeValueByIndex} index={i} value={v}  />
      ))}
   </>
)
```

```
const Item = React.memo(({ onChangeByIndex, index, value }) => {
    const handleChange = useMemo(() => onChangeByIndex(index), [onChangeByIndex, index]);

    return (
        <Input value={value} onChange={handleChange} />
    )
}) 
```

This way we get benefits for memoizing the handleChangeValueByIndex which passe to `Item`. The parent component rerender won't affect the child `Item`.

**Comment by gaearon** [comment]
In this case the “event” is the inner curried part. In general I would not recommend currying React event handlers. It’s a pretty confusing pattern to read and deal with.

**Comment by davidcoleman007** [comment]
> The `useEvent` proposal definitely goes in the right direction, as the misuse of `useCallback` is something we've all seen countless times across teams and organizations.
> 
> I do want to express some strong concerns around the naming as `useEvent` could potentially be very confusing for less experienced engineers or those who are coming to React from other frameworks. The dangerous naming correlation to DOM events makes it quite misleading at first glance and seems to hide its true purpose.
> 
> I much prefer `useStableCallback` as someone else suggested, or `useHandler` as I feel it makes it easier to document and explain its usage moving forward.

As a long time user of react, I also agree. `useEvent` makes me assume we are using an emitter and i would avoid handler for the same reason as it implies action in response to an emitted event.  I like your suggestion of `useStableCallback` over `useEvent`

**Comment by zheeeng** [comment]
> In this case the “event” is the inner curried part. In general I would not recommend currying React event handlers. It’s a pretty confusing pattern to read and deal with.

Can we have documented rules or principles to note this recommendation? Users can't avoid getting trapped by using `useEvent` in ways out of the `When useEvent should not be used` chapter listed.

**Comment by nikparo** [comment]
> > In this case the “event” is the inner curried part. In general I would not recommend currying React event handlers. It’s a pretty confusing pattern to read and deal with.
> 
> Can we have documented rules or principles to note this recommendation? Users can't avoid getting trapped by using `useEvent` in ways out of the `When useEvent should not be used` chapter listed.

The RFC states that the event created by `useEvent` will throw if called during render, which is what you did with your curried handler.

**Comment by sompylasar** [comment]
One more argument against the "useEvent" naming. The following statement derived from the RFC doesn't make sense in English:

> [The RFC states that] the event created by useEvent will throw if called

How can an event throw? How can an event be called? An event is something that happens. Sometimes we say "an event" as a shorthand to "an event object" that not only mentions what happened, but also describes the details about what happened. 

But it feels unintuitive to use "an event" as a shorthand for "an event handler function" that processes what happened during an event.

While writing the above, I think I understood one way of thinking where these can be equivalent: if we treat an event as something not instantaneous that has duration (which is not how events are usually treated in software engineering), and then treat the side effects as an unseparable part of an event, then the function that processes the event is the part of the event, hence "using an event" makes some sense.

**Comment by nikparo** [comment]
> How can an event throw? How can an event be called?

@sompylasar I think that was just me accidentally leaving out "function" from "event function". Sorry about that.

**Comment by sompylasar** [comment]
@nikparo Nothing to be sorry for, I think that's important and that proves my point. We use programming languages and naming conventions to communicate, but this particular (useEvent) naming introduces a lot of space for ambiguity. By the way, "event function" also doesn't make a lot of sense in the context of "an event" as a marker for a point in time when something happened instantaneously.

**Comment by alirezamirian** [comment]
[the rfc](https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md) sounds nice. In almost any react app, I end up using a `useLatest` hook to ref-ify non-reactive values I only need in response to some event. I wonder if all usages of `useLatest` are related to event handling and hence covered by the proposed hook for creating event handlers.

On a different note, the name `useEvent` confused me in the beginning and I see many people are also talking about it as well.
🟢 `useState` => returns state (and setter)
🟢 `useRef` => returns ref
🟢 `useCallback` => returns callback
🟢 `useMemo` => returns memorized value
🔴 `useEvent` => returns event **handler**

I get that [having a short name](https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md#:~:text=Having%20a%20short%20name%20helps) is nice, but confusing "event" and "event handler" is a much more considerable drawback IMO.
I think `useEventHandler` or `useHandler` (if having a short name is deemed very important) are better choices.

**Comment by vadzim** [comment]
yeah, it's implemented in https://www.npmjs.com/package/react-use-handler
would be nice to have it in react.

**Comment by zheeeng** [comment]
> > > In this case the “event” is the inner curried part. In general I would not recommend currying React event handlers. It’s a pretty confusing pattern to read and deal with.
> > 
> > 
> > Can we have documented rules or principles to note this recommendation? Users can't avoid getting trapped by using `useEvent` in ways out of the `When useEvent should not be used` chapter listed.
> 
> The RFC states that the event function created by `useEvent` will throw if called during render, which is what you did with your curried handler.

`useEvent` shouldn't be used in render-props, a common react pattern. it's explainable but it is hard to realize intuitively. Should we list more cases in RFC?

**Comment by BirdTho** [comment]
Gentlemen: `useGranularCallback` from `npm i granular-hooks`
BTW, I use their useGranularEffect() a lot because I don't want useEffect style triggering when only certain vars should trigger updates. It takes two lists of arguments, primary and secondary deps as arrays. primary deps trigger calling the hook or changing the callback, secondaries don't.

**Comment by nikparo** [comment]
@BirdTho I checked the code of [granular-hooks](https://github.com/gfox1984/granular-hooks). The whole library is just a roundabout way of calling hooks while leaving out some dependencies. The secondary dependency array does nothing at all - it doesn't matter whether you include it or not. It does not protect you against stale values!

If you want a hook that doesn't nag at you about `exhaustive-deps`, then you are better off with `export const useGranularCallback = useCallback;`.

**Comment by BirdTho** [comment]
I disagree.
Secondary deps array does something:
https://github.com/gfox1984/granular-hooks/blob/8fbb19cbb949e5cd7f15827c4a6a6a60c2465788/src/hooks/useGranularHook.ts#L17

**Comment by BirdTho** [comment]
@Hypnosphi So basically, if I'm not going to use concurrent mode, this is fine then?

**Comment by nikparo** [comment]
> It is clearly being included in what's passed to the callback, along with the primary deps.
It's only running the equivalency check on the primary deps.

@BirdTho They are being passed to the hook* (not callback), sure, but they are only updated when the primary deps are updated. There is no magic connection between the values in the closure and the values in the dependency array. In other words, the callback closure remains stale until the primary deps change.

But this is getting off topic, so if you don't believe me I suggest you simply test it yourself with and without the so called secondary deps.

**Comment by Moranilt** [comment]
Seems like it's useless with onChange events for input:
```tsx
type InputPropsA = {
  name: string;
  value: string;
  setValue: (value: string) => void;
};

const InputA: React.FC<InputPropsA> = ({ name, value, setValue }) => {
  const onChangeHandler = useEvent((e: React.ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
  });

  return <input name={name} onChange={onChangeHandler} value={value} />;
};

const Form = () => {
  const [firstname, setFirstname] = useState('');
  const [lastname, setLastname] = useState('');
  const [middlename, setMiddlename] = useState('');

  return (
    <form>
      <InputA name="firstname" setValue={setFirstname} value={firstname} />
      <InputA name="lastname" setValue={setLastname} value={lastname} />
      <InputA name="middlename" setValue={setMiddlename} value={middlename} />
    </form>
  );
};
```

Even if you put every setState into useCallback - it's not working for me right now with [this hook](https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md#internal-implementation)

**Comment by artem-malko** [comment]
@Moranilt 

The main idea is to memoize a callback, which is passed to another component. In your example you're trying to memoize a callback, which is passed into `<input />`. But there is no point to do it, cause, `<input />` does have any instruments to use that memoized callback.

So, in your example you could memoize set*** callback. But in your case they are memoized already) 

But, I have a fix for your example:

```jsx

// You have to use memo here, to make memoization work
const Input = memo<InputPropsA>(({ name, value, setValue, onClick }) => {
  // You do not need any memoization here
  return <input name={name} onChange={(e) => e.taget.value} value={value} onClick={onClick} />;
});

const Form = () => {
  const [firstname, setFirstname] = useState('');
  const [lastname, setLastname] = useState('');
  const [middlename, setMiddlename] = useState('');
  
  // And this is a correct memoization, cause onInputClick is passed into Input
  const onInputClick = useEvent(() => {
    console.log(firstname, lastname, middlename);
  });

  return (
    <form>
      <Input name="firstname" setValue={setFirstname} value={firstname} onClick={onInputClick} />
      <Input name="lastname" setValue={setLastname} value={lastname} onClick={onInputClick} />
      <Input name="middlename" setValue={setMiddlename} value={middlename} onClick={onInputClick} />
    </form>
  );
};
```

**Comment by kambing86** [comment]
to be honest, I'm not sure why the implementation of the useEvent has to be so complicated, why not just this?

```js
function useEvent(callback) {
  const callbackRef = useRef(callback)
  callbackRef.current = callback
  
  return useCallback((...args) => {
    callbackRef.current(...args)
  }, [])
}
```

the only issue with this is when using typescript, it's hard to get the definition, maybe if there is any typescript expert can help to solve this
```ts
function useEvent(callback: (...args: any[]) => void) {
  const callbackRef = useRef(callback)
  callbackRef.current = callback

  return useCallback((...args: any[]) => {
    callbackRef.current(...args)
  }, [])
}
```

with this implementation, there is no such thing as when will the current is switched, because it will be switched every single time, and it can be used in rendering as well

**Comment by jampy** [comment]
@kambing86 this will not work properly in "concurrent mode", see https://github.com/facebook/react/issues/14099#issuecomment-440794435

**Comment by sarakusha** [comment]
```ts
function getStateAsync(setter) {
  return new Promise(resolve => {
     setter(prevValue => {
        resolve(prevValue);
        return prevValue; // The value has not changed.
     });
  })
}

function Chat() {
  const [text, setText] = useState('');

  const onClick = useCallback(async () => {
    sendMessage(await getStateAsync(setText));
  }, []); // There is no dependency array.

  return <SendButton onClick={onClick} />;
}
```
This [gist](https://gist.github.com/sarakusha/2fc9d1d4d3e732d3ac890993a9c80a4a) is written in typescript and allows you to get several states at once.

**Comment by kfcaio** [comment]
While it's not fully available, what could we use to achieve the closest behavior? https://github.com/vadzim/react-use-handler, https://github.com/gfox1984/granular-hooks, or some gist above? I'm using React 16

**Comment by KingSora** [comment]
> @sokra An alternate would be:
> 
> ```js
> function useEventCallback(fn) {
>   let ref = useRef();
>   useLayoutEffect(() => {
>     ref.current = fn;
>   });
>   return useCallback(() => (0, ref.current)(), []);
> }
> ```
> 
> This doesn't require the args like yours has. But again, you can't call this in the render phase and the use of mutation is dicey for concurrent.

@sophiebits wouldn't `useImperativeHandle` be even closer to the desired behavior:
```js
function useEventCallback(fn) {
  const ref = useRef(fn);
  useImperativeHandle(ref, () => fn);
  return useCallback((...args) => ref.current?.(...args), []);
}
```

**Comment by HansBrende** [comment]
I think we can save some memory here by writing it this way (avoiding what seems to be an unnecessary `useCallback` call in the above proposals):

```javascript
export const useEvent = (fn) => {
  const ref = useRef([fn, (...args) => ref[0](...args)]).current;
  useLayoutEffect(() => {
    ref[0] = fn;
  });
  return ref[1];
}
```

**Comment by jewalky** [comment]
I bumped into this thread while already having a solution and trying to understand if I'm reinventing the wheel :) 

Looks like I'm not, as my solution to the problem is different than existing options:

```typescript
export default function useConstCallback<T extends CallableFunction>(func: T, deps?: React.DependencyList): T {
  const ref = useRef<T>(func)

  useMemo(() => {
    ref.current = func
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps)

  return useCallback((...params) => ref.current(...params), []) as unknown as T
}
```

The main difference compared to `useEvent` above is that I'm abusing `useMemo()` (which runs instantly during hook call) instead of using `useLayoutEffect()` or `useEffect()` which calls only after the render;

Considering `useCallback(x, deps)` is functionally the same as `useMemo(() => x, deps)` (see implementation in React: https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberHooks.js#L2242), this works exactly like `useCallback()` would, with the exception of having a constant reference.

**Comment by BirdTho** [comment]
Stuff like this is why Vue or Angular are still alive.

**Comment by JoaquimEsteves** [comment]
Given that the 'useEvent' hook has [gone the way of the doodoo](https://github.com/facebook/react/pull/25881) what are the plans to solve this issue?

I was a big fan of the 'useEvent' hook - and end up having to copy my own implementation in every other project (something like [use-event-callback](https://github.com/Volune/use-event-callback) or the one @HansBrende mentioned in that issue).

To bring this issue to a final close why not encourage a simply "hook recipe" on the documentation?

Something like:

```
## Avoiding Invalidated Memoized Objects

Slap your function into a ref that's updated through useLayoutEffect.

<recipe here>

Caveats:

bla bla SSR bla bla
```

I understand that the team is currently investigating other options for the 'rendering optimizations are cumbersome to type' problem- but us every day developers don't have access to these tools.

Indeed, should they be released we might not be able to use them because we're stuck on an old version of React for some reason or another.

**Comment by vzaidman** [comment]
> ```
>       setB(b => {
>         callApi(a,b)
>         return b
>       })
> ```

Is this the correct code? To me it reads like-
"On mount, only once, `callApi` with the latest a and b values from the useState hook."

**Comment by troglotit** [comment]
@vzaidman you're reading correctly. But, I have to concede, the code is quite abstracted.

**Comment by Effanuel** [comment]
> Had a quite interaction heavy app, so had to optimize rendering quite heavily. So leaned into `setState(updaterFn)`
> 
> ```js
> function App() {
>   let [a, setA] = useState()
>   let [b, setB] = useState()
>   
>   useEffect(() => {
>     setA(a => {
>       setB(b => {
>         callApi(a,b)
>         return b
>       })
>       return a
>     })
>   }, [])
> 
>   // which fundamentally can be changed to either adding custom hook, 
>   // or extending React's useEffect/useMemo/useCallback via RFC
>   useStateEffect((a,b) => {
>     callApi(a,b)
>   }, [], [setA, setB])
> }
> ```
> 
> Which, I think, precisely solves all the problems: optimize useCallback, premature "useEvent all the things". But requires passing setStates instead of state itself as props.
> 
> Also, not sure about this concern in another issue, maybe I'm missing something:
> 
> > You can useReducer or setState(updaterFn) to avoid closing over the variables. This doesn't solve all use cases (especially when you also need to perform side effects) but gets you pretty close.

Can you make this code more complex, its too simple

**Comment by nvmnghia** [comment]
> Had a quite interaction heavy app, so had to optimize rendering quite heavily. So leaned into `setState(updaterFn)`
> 
> ```js
> function App() {
>   let [a, setA] = useState()
>   let [b, setB] = useState()
>   
>   useEffect(() => {
>     setA(a => {
>       setB(b => {
>         callApi(a,b)
>         return b
>       })
>       return a
>     })
>   }, [])
> 
>   // which fundamentally can be changed to either adding custom hook, 
>   // or extending React's useEffect/useMemo/useCallback via RFC
>   useStateEffect((a,b) => {
>     callApi(a,b)
>   }, [], [setA, setB])
> }
> ```
> 
> Which, I think, precisely solves all the problems: optimize useCallback, premature "useEvent all the things". But requires passing setStates instead of state itself as props.
> 
> Also, not sure about this concern in another issue, maybe I'm missing something:
> 
> > You can useReducer or setState(updaterFn) to avoid closing over the variables. This doesn't solve all use cases (especially when you also need to perform side effects) but gets you pretty close.

Looks really weird. Why didn't you just use `a` & `b`, leave the deps array empty then add some lint ignore & comment?

**Comment by tibbe** [comment]
> @sokra An alternate would be:
> 
> ```js
> function useEventCallback(fn) {
>   let ref = useRef();
>   useLayoutEffect(() => {
>     ref.current = fn;
>   });
>   return useCallback(() => (0, ref.current)(), []);
> }
> ```
> 
> This doesn't require the args like yours has. But again, you can't call this in the render phase and the use of mutation is dicey for concurrent.

Perhaps a stupid question. What is the `0` doing there? Why isnt' it just `ref.current()`?

**Comment by vadzim** [comment]
> What is the `0` doing there? Why isnt' it just `ref.current()`?

It's a trick to call a function without setting `this` and an equivalent of that:
```ts
const f = ref.current;
f()
```

**Comment by guoyunhe** [comment]
Currently there are many community implementation of this hook:

- [use-latest](https://www.npmjs.com/package/use-latest) from storybook
- [use-latest-callback](https://www.npmjs.com/package/use-latest-callback)
- [use-callback-stable](https://www.npmjs.com/package/use-callback-stable)

**Comment by dmytro-ulianov** [comment]
well guys, i think we finally have a solution and this issue can be closed 🎉

long overdue solution → [useEffectEvent – React](https://react.dev/reference/react/useEffectEvent)

update: looks like it's different from what was discussed here, and is created specifically for useEffect usage 🫠

**Comment by mgmolisani** [comment]
It feels extremely crazy that in 7 years we are no closer to a solution here yet we've seen heaps of changes that many applications in production code bases won't even adopt. We had mutable `this.props` in class components and now there is no fool-proof replacement. There seems to always be a footgun preventing the function from being used in some context with the argument it's protecting developers from worse evils. I don't disagree that `useCallback` does prevent these errors but theres times where not having an alternate pattern just simply blocks development.

It seems like the closest we have gotten is roughly the `useEventCallback` mentioned many times cloning to a ref and then updating in a layout effect. This works for many cases but it has been mentioned that there is a pitfall due to the order of layout effects running in children first. This makes it stale if run in a deeper layout effect.

```ts
function useEventCallback(fn) {
  const ref = useRef(fn);
  useLayoutEffect(() => {
    ref.current = fn;
  });
  return useCallback((...args) => ref.current(...args), []);
}
```

Additionally, as it is not a React built-in, you still have to declare the result in dependency arrays which ruins the semantics of the effect as this really is just a mutable ref and not reactive. Whether it will ever trigger the effect or not isn't relevant as by declaring it as a dependency, you've told the reader that the function inside should fire if it ever changes which is usually not what you want. The (slow moving) `useEffectEvent` intends to solve that but its also still unclear if that can be used for synchronous event emission in a `useLayoutEffect`.

What I don't see a lot of people talk about in these types of threads is how the browser works natively. When creating React components, I almost always call back to browser behaviors. Whether it is simple like naming props like `isHidden`, `notVisible`, or `hidden`, or thinking about what the shape and trigger timing of events, it seems like the browser usually has an analogous solution. 

For example, let's think about browser events like `focusout`. It is obvious to us that this will trigger synchronously when we click away from an input. But what about the case where the property `disabled` becomes true?  This also triggers a synchronous `focusout` event. One could correlate this operation in React as a prop change triggering an event in the child (blasphemy!). How would we implement something like this in React?

Interestingly, the behavior in React for this synchronous `focusout` event case is _always_ stale. If React changes the `disabled` property, the callback invoked will be from the previous render as React cannot intervene to replace the callback before it's processed. Here's a [sandbox](https://codesandbox.io/p/sandbox/zm5385?file=%2Fsrc%2FApp.js) to prove it. Additionally, you can try `useInsertionEffect` as well and the issue remains as React cannot circumvent the synchronous nature of the `disabled` setter triggering `focusout` synchronously.

Say we have a complex state like `const active = hovered || focused || x > 5` and `x` comes from props. Until we determine the value of `active`, we don't know what to emit in some `onActiveChange` event. Had `x` not been in the equation, maybe we could do it in the handlers for `onMouseEnter/Leave` and `onFocus/Blur` precomputing (and repeating) the final derived state before emitting. But given `x` can change, much like `disabled` for inputs, we have a potential event we would want to handle in an effect. To act like the synchronous `focusout` event for inputs, we would want to use `useLayoutEffect` here to tell other parts of the tree about the state update. Now, again, were back to that caveat, though, the sandbox disproves that it deviates from regular React behavior.

Some might be asking at this point, "why not just lift state up?" and execute a callback wherever we are setting `x`. Yes, "derived state from props"  bad... I know, I know. I'm exhausted hearing this solution for so many problems in React. 

Possible issues include:
* `x` is way way above our component
  * You really want me to also lift up the `hover` and `focus` values? And then also compute the derived state to emit the event. Should I put all the application state in the `App.tsx` file? Then possibly pass all three of those values way back down to my component? What if my component is also in an array? This is an encapsulation nightmare. Sometimes _uncontrolled_ components are actually cleaner.
* `x` might come from a library that only provides `useState` style hook, no events
  * Theres no event, looks like we are back to using `useLayoutEffect` anyway.
  * If it does provide handlers that correlate to the state, is it really worth it to recompute the state to control the state separately in callbacks?
* Maybe `hovered` and/or `focused` can't be lifted
  * They might come from hooks like `useHover` and `useFocus` that return a state. Lifting these might be impossible due to loop restrictions.

While I'm going to continue the verbose process of copying the code above to avoid lint errors and semantic violations in my dep array, it would be infinitely better if we just had some simple API to unpack a reactive value into some stable flavor like we had with class components in callbacks. Before, the inverse was as easy as `const { prop } = this.props;`. Certainly, there is some `useStable(value)` that can save us and be more compatible with the linting tools.

**Comment by andres-kovalev** [comment]
> [@muqg](https://github.com/muqg) In concurrent mode, last render doesn't necessarily mean "latest committed state". So a low-priority render with new props or state would overwrite a reference used by current event handler.

Probably I misunderstood your comment, but from the thread context I guess by "low-priority render" you mean a render of the same component. But why would we perform an older (unnecessary) render for a component if we already have a result of a newer render?

**Comment by sophiebits** [comment]
useEventEffect has now been released and has similar functionality to the useEventCallback function mentioned several times in this thread: https://react.dev/blog/2025/10/01/react-19-2#use-effect-event. With useEventEffect you don't need to (and shouldn't) list the function in the dependency array, but you're not allowed to pass it to a deeper component; instead, the deeper component should wrap the callback if it needs to do so.

For functions called during the render process, useEventEffect is not applicable because the identity of the function has semantic meaning. For those it's desired that useCallback changes its identity when the dependencies change (otherwise the deeper component would have no idea when it needs to re-call the function) but if you have performance issues then use of the profiler, useMemo, and/or the React Compiler can fix most issues.

I'll close this issue now as the team is happy with the launch of useEventEffect, and I don't anticipate more work in this direction in the foreseeable future.

**Comment by avasuro** [comment]
Thanks a lot for the release - great to see progress in this direction!
The new `useEffectEvent` already solves about half of the original issue - especially the part about stable dependencies in `useEffect` (we finally have a function that doesn’t need to go into the deps array).

However, the other half of the problem remains because of the current limitation:
> “You’re not allowed to pass it to a deeper component.”

This creates a few issues:
- When a function is used **both** in `useEffect` **and** passed to a child, we must keep a `useCallback` approach and create a duplicate via `useEffectEvent`, adding unnecessary boilerplate and extra level of function wrapping.
- Since the `useCallback` result is passed to the child, it still changes when dependencies change, causing unnecessary re-renders / re-calculations of memoized children.

<details>
  <summary>Example</summary>

  ```ts
  // Some function that is imperative by its nature
  // (e.g., an event handler / setter / action).
  // I’d love to use `useEventCallback` here,
  // but I can’t - since it’s also passed to a child component.
  const doSomething = useCallback(() => {
    // ...do something
  }, [someDep1, someDep2]);

  // ...and I still have to create a “duplicate”
  // to have a stable version for useEffect:
  const doSomethingStable = useEffectEvent(doSomething);
  useEffect(() => {
    console.log(someDep); // do something with dependency
    doSomethingStable(); // and call our function
  }, [someDep]);

  // But in the child prop, I’m still forced to use
  // the "unstable" version of this function,
  // which "breaks" memoization: 
  return <MemoizedChild onSomething={doSomething} />;
  ```
</details>

Is there any specific reason why "You’re not allowed to pass it to a deeper component" limitation existst?

P.S.: Oh, actually I didn’t notice at first, but the official docs also discourage using it inside other hooks (except useEffect and useLayoutEffect).
That makes this solution even less applicable to this case - we still don’t have a proper, official way to define a truly “stable” callback.

**Comment by sophiebits** [comment]
This was deliberate because in general you can't know if the child component cares about the identity of the function or not. The child might want to re-call if the function would give a different value, but it can't do that if you wrap in useEventCallback in the caller. I know it's subtle but making sure you can reason about each component's correctness independently is a core part of what we look for when designing React features.

(Forgot to close last time, closing now!)

**Comment by gaearon** [comment]
> Oh, actually I didn’t notice at first, but the official docs also discourage using it inside other hooks.

This doesn't sound right to me. The limitations are listed [here](https://react.dev/learn/separating-events-from-effects#limitations-of-effect-events).

You *can* use it inside other Hooks, as long as it's next to the Effect that relies on it.

See https://github.com/reactjs/react.dev/issues/8075#issuecomment-3397067889 for example.

**Comment by avasuro** [comment]
@gaearon , maybe I’m looking in the wrong place, but when I follow the link you provided (“Limitations of effect events”), there’s a big bold note that says:
>  - Only call them from inside Effects.
> - Never pass them to other components **or Hooks**.

The example right below that text also mentions that using them inside other hooks is discouraged:
> useTimer(onTick, 1000); // 🔴 Avoid: Passing Effect Events

So, probably, I should clarify my message: it was meant to say “...discourage using it inside other hooks (_except useEffect and useLayoutEffect_)”.

**Comment by gaearon** [comment]
Right, I thought you meant defining them inside custom Hooks (which you can do).

Indeed you can’t call them from non-Effects, which as @sophiebits described is intentional. Conceptually I like to think of an Effect Event as being a “part of” its Effect’s code, but written outside of it to escape reactivity.
